[Improvement] notifications - pixelfed

eufelipemateus opened the pull request:
# Notification System Improvements – Part 1 of 2

This PR aims to improve how notifications are handled by introducing the use of the `read_at` field in the notification list and highlighting unread notifications. This is **part 1 of a 2-part update**.

## Changes Included in This PR

- Updated the front-end to visually highlight unread notifications.
- Created three new API routes:
  - Mark a notification as read.
  - Mark a notification as unread.
  - Get the total count of unread notifications.
- Added a scheduled cron job that runs every 6 hours to delete notifications that have been marked as read for over 3 months.

eufelipemateus committed:
change front

eufelipemateus committed:
add mark and unmark notification

eufelipemateus committed:
add mark all as read

eufelipemateus committed:
return  total unread

eufelipemateus committed:
remove space

eufelipemateus committed:
remove old notifications

dansup commented on app/Http/Controllers/Api/ApiV1Controller.php:
```diff
@@ -4494,4 +4494,95 @@ public function statusUnpin(Request $request, $id)
+     */
+    public function accountNotificationsMarkAsRead(Request $request)
+    {
+        /* abort_if(! $request->user(), 403);
```
Why is this commented out?

eufelipemateus replied:
This was used to make testing easier, it wasn't supposed to be in PR, I'm going to remove the comment

eufelipemateus replied:
fixed in: b390415f99d04a4a2510c990062bfa54a117f9af

dansup commented on app/Http/Controllers/Api/ApiV1Controller.php:
```diff
@@ -4494,4 +4494,95 @@ public function statusUnpin(Request $request, $id)
+        }
+
+        $notification = Notification::whereProfileId($pid)->findOrFail($id);
+        $notification->read_at = 'now()';
```
This should be

```
$notification->read_at = now();
```

eufelipemateus replied:
fixed in: b390415f99d04a4a2510c990062bfa54a117f9af

dansup commented on app/Http/Controllers/Api/ApiV1Controller.php:
```diff
@@ -4494,4 +4494,95 @@ public function statusUnpin(Request $request, $id)
     }
+
+    /**
+     * GET /api/v1/notifications
```
This needs to be updated

eufelipemateus replied:
fixed in: b390415f99d04a4a2510c990062bfa54a117f9af

dansup commented on app/Http/Controllers/Api/ApiV1Controller.php:
```diff
@@ -4494,4 +4494,95 @@ public function statusUnpin(Request $request, $id)
+        $pid = $user->profile_id;
+        $id = $request->input('id');
+
+        if (empty($id)) {
```
This needs to be in [another method](https://docs.joinmastodon.org/methods/notifications/#clear) to follow the Mastodon API

eufelipemateus replied:
OKay. I will apply

eufelipemateus replied:
fixed in: b390415f99d04a4a2510c990062bfa54a117f9af

dansup commented on app/Http/Controllers/Api/ApiV1Controller.php:
```diff
@@ -4494,4 +4494,95 @@ public function statusUnpin(Request $request, $id)
+    public function accountNotificationsMarkAsUnread(Request $request)
+    {
+        /*
+        abort_if(! $request->user(), 403);
```
This shouldn't be commented out

eufelipemateus replied:
fixed in: b390415f99d04a4a2510c990062bfa54a117f9af

dansup commented on app/Http/Controllers/Api/ApiV1Controller.php:
```diff
@@ -4494,4 +4494,95 @@ public function statusUnpin(Request $request, $id)
+    }
+
+    /**
+     * POST /api/v1/notifications/mark_as_unread
```
What is the point of this? I don't think we need to support marking a notification as unread. Please remove this.

eufelipemateus replied:
removed in: b69c95748b1e98d10b72eb7bb458c1020a2b228e

dansup commented on app/Http/Controllers/Api/ApiV1Controller.php:
```diff
@@ -4494,4 +4494,95 @@ public function statusUnpin(Request $request, $id)
+    }
+
+    /**
+     * GET /api/v1/notifications/status
```
Please remove this

eufelipemateus replied:
A change to /api/v1/notifications/unread_count like mastodon spec in: b390415f99d04a4a2510c990062bfa54a117f9af

dansup commented on routes/api.php:
```diff
@@ -135,6 +135,10 @@
         Route::put('media/{id}', 'Api\ApiV1Controller@mediaUpdate')->middleware($middleware);
         Route::get('mutes', 'Api\ApiV1Controller@accountMutes')->middleware($middleware);
         Route::get('notifications', 'Api\ApiV1Controller@accountNotifications')->middleware($middleware);
+        Route::post('notifications/mark_as_read', 'Api\ApiV1Controller@accountNotificationsMarkAsRead')->middleware($middleware);
```
These do not follow the Mastodon API spec, and need to be updated.

eufelipemateus replied:
fixed in: b390415f99d04a4a2510c990062bfa54a117f9af

dansup reviewed (changes requested):
I think we should adopt a more efficient strategy of using notification read markers, similar to Mastodon

https://docs.joinmastodon.org/methods/notifications/#unread-count

eufelipemateus commented:
I will refactor this PR following the Mastodon specification: https://docs.joinmastodon.org/methods/notifications

eufelipemateus committed:
change to mastodon specs

dansup commented:
> I will refactor this PR following the Mastodon specification: https://docs.joinmastodon.org/methods/notifications

Perfect, thank you!

eufelipemateus committed:
fix bugs

eufelipemateus committed:
Apply adjust to read notifications

eufelipemateus committed:
Apply adjust to read notifications

eufelipemateus committed:
remove markUnread function from front

eufelipemateus commented:
Good afternoon! I’ve refactored this PR according to the recommendations you provided. Below is a summary of the changes made:

### 🔄 Backend changes (API routes updated to follow Mastodon specs):

- The route `POST /api/v1/notifications` was replaced with  
  `POST /api/v1/notifications/{id}/dismiss`, which is now used to **mark a notification as read**.

- The route `POST /api/v1/notifications/mark_as_unread`, which was used to **mark a notification as unread**, has been **removed**.

- A new route `POST /api/v1/notifications/clear` was created to **mark all notifications as read**.

- The route `GET /api/v1/notifications/status` was renamed to  
  `GET /api/v1/notifications/unread_count`, and its response was simplified to return only the `count` field, reflecting the total unread notifications depending on the input query parameters.

### 💻 Frontend changes:

- Removed the **“mark as unread”** button, since the related backend functionality no longer exists.

Let me know if you need any further changes or adjustments!
